DECKGL integration - Phase 1

mistercrunch opened the pull request:
![deck](https://user-images.githubusercontent.com/487433/32382851-9404707a-c073-11e7-9546-69c7224fb3cf.gif)
<img width="761" alt="screen shot 2017-11-03 at 9 28 58 am" src="https://user-images.githubusercontent.com/487433/32384867-80691164-c079-11e7-82ed-ab338b2b6e0d.png">
Adding a new set of geospatial visualizations building on top of the
awesome deck.gl library. https://github.com/uber/deck.gl

While the end goal it to expose all types of layers and let users bind
their data to control most props exposed by the deck.gl API, this
PR focusses on a first set of visualizations and props:

* ScatterLayer
* HexagonLayer
* GridLayer
* ScreenGridLayer

This is a large PR, so here's an overview of what is in here:
* a common `DeckGLContainer` that we'll use to render all deck related visualizations
* 4 new vizes
* an example dash loaded as part of `superset load_examples`
* a new `FixedOrMetricControl` allowing to provide a fixed value or binding to a metric
* ...

mistercrunch committed:
DECKGL integration

Adding a new set of geospatial visualizations building on top of the
awesome deck.gl library. https://github.com/uber/deck.gl

While the end goal it to expose all types of layers and let users bind
their data to control most props exposed by the deck.gl API, this
PR focusses on a first set of visualizations and props:

* ScatterLayer
* HexagonLayer
* GridLayer
* ScreenGridLayer

xrmx commented on superset/cli.py:
```diff
@@ -133,9 +133,15 @@ def load_examples(load_test_data):
     print("Loading [Misc Charts] dashboard")
     data.load_misc_dashboard()
 
+    print("Loading DECK.gl demo")
```
merge resolution mismatch?

mistercrunch replied:
still WiP, I'll ping when ready for review...

xrmx commented on package.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Is this in the right place?

xrmx commented on superset/viz.py:
```diff
@@ -1731,6 +1689,112 @@ def get_data(self, df):
 
+class BaseDeckGLViz(BaseViz):
+
+    """Rich maps made with Mapbox"""
```
c&p desription :)

fabianmenges commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,118 @@
+  constructor(props) {
+    super(props);
+    this.onChange = this.onChange.bind(this);
+    const type = props.value ? props.value.type : props.default.type || 'fix';
```
'fix' should probably be a global constant that you reference everywhere. e.g. declare `VALUE_TYPE_FIX = 'fix'` and use that everywhere. In case you ever want to refactor, this will make it a lot easier.

fabianmenges replied:
I think adding parenthesis will make this line easier to read:
```
const type = props.value ? props.value.type : (props.default.type || 'fix');
vs
const type = (props.value ? props.value.type : props.default.type) || 'fix';
```
I always get confused about the order of operations in these cases, but maybe thats just me.
another option is to pull out `if (props.value)`

mistercrunch replied:
yes like an enum, I'm going with
```
const controlTypes = {
  fixed: 'fix',
  metric: 'metric',
};
```

fabianmenges commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,118 @@
+    this.state = {
+      type,
+      fixedValue: type === 'fix' ? value : '',
+      metricValue: type === 'metric' ? value : null,
```
`VALUE_TYPE_METRIC` instead of `metric`? Also it looks weird that one of them  defaults to `''` while the other one defaults to `null`

mistercrunch replied:
one is a string (TextControl) and the other is an unselected SelectControl...

fabianmenges commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,118 @@
+          <PopoverSection
+            title="Fixed"
+            isSelected={type === 'fix'}
+            onSelect={this.onChange.bind(this, 'fix')}
```
is there a reason you bind `this.onChange` here? Afaik calling `bind` once in the constructor is recommended and more performant (https://reactjs.org/docs/handling-events.html) 

> We generally recommend binding in the constructor or using the class fields syntax, to avoid this sort of performance problem.

mistercrunch replied:
Well here I'm binding with 'fix' as well, I guess I could bind to this on the constructor and re-bind that to 'fix' here.

fabianmenges commented on superset/assets/javascripts/explore/components/controls/ViewportControl.jsx:
```diff
@@ -0,0 +1,97 @@
+  onChange: () => {},
+  default: { type: 'fix', value: 5 },
+  value: {
+    longitude: 6.85236157047845,
```
Could you put in a comment what these coordinate are?

fabianmenges commented on superset/assets/javascripts/explore/stores/controls.jsx:
```diff
@@ -424,6 +432,12 @@ export const controls = {
   },
 
   groupby: groupByControl,
+  dimension: Object.assign({}, groupByControl, {
```
Further up you are doing 
```
{
       ...this.props.value,
       [ctrl]: value,
}
```
here you are using `Object.assign()` instead of 
```
{
   ... groupByControl,
  label: t('Dimension'),
  description: t('Select a dimension'),
  multi: false,
  default: null,
}
```

mistercrunch replied:
yeah let's move towards spread operator instead of `Object.assign`, we should eventually add a linting rule for this

fabianmenges commented on superset/assets/javascripts/explore/stores/controls.jsx:
```diff
@@ -1317,6 +1382,20 @@ export const controls = {
+    renderTrigger: true,
+    description: t('Parameters related to the view and perspective on the map'),
+    default: {
+      longitude: 6.85236157047845,
```
These looks like the same numbers that are hardcoded as defaults, do we need them in both locations?

mistercrunch replied:
moving as `defaultViewport` in utils/geo.js

mistercrunch committed:
Addressing comments

mistercrunch commented:
Thanks for the review @fabianmenges ! I addressed all the comments and improved a few other things along the way.

mistercrunch committed:
lint

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+  value: PropTypes.object,
+  isFloat: PropTypes.bool,
+  datasource: PropTypes.object,
+  default: PropTypes.object,
```
should this be a shape?

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+    });
+  }
+  setType(type) {
+    this.setState({ type }, this.onChange);
```
`setState` should be a function now https://reactjs.org/docs/react-component.html#setstate

mistercrunch replied:
oh interesting I had not seen that change

mistercrunch replied:
After reviewing the updated React docs it doesn't sound like the `updater` as first arg is a preferred way of calling `setState` or that passing an object will get deprecated in the future of anything. Any other rational to change thing over? I'd rather not go and change hundreds of setState calls in the code base to add unneeded brackets...

mistercrunch replied:
I tried briefly to find a related eslint linting rule that would explain why an `updater` function is preferred and oculdn't find any...

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+    this.setState({ type }, this.onChange);
+  }
+  setFixedValue(fixedValue) {
+    this.setState({ fixedValue }, this.onChange);
```
ditto functional `setState`

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+    this.setState({ fixedValue }, this.onChange);
+  }
+  setMetric(metricValue) {
+    this.setState({ metricValue }, this.onChange);
```
ditto functional `setState`

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+          >
+            <TextControl
+              isFloat
+              onChange={this.setFixedValue.bind(this)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+          <PopoverSection
+            title="Fixed"
+            isSelected={type === controlTypes.fixed}
+            onSelect={this.onChange.bind(this, controlTypes.fixed)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+            <TextControl
+              isFloat
+              onChange={this.setFixedValue.bind(this)}
+              onFocus={this.setType.bind(this, controlTypes.fixed)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+          <PopoverSection
+            title="Based on a metric"
+            isSelected={type === controlTypes.metric}
+            onSelect={this.onChange.bind(this, controlTypes.metric)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+              name="metric"
+              options={metrics}
+              onFocus={this.setType.bind(this, controlTypes.metric)}
+              onChange={this.setMetric.bind(this)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/FixedOrMetricControl.jsx:
```diff
@@ -0,0 +1,124 @@
+              {...controls.metric}
+              name="metric"
+              options={metrics}
+              onFocus={this.setType.bind(this, controlTypes.metric)}
```
you shouldn't bind in a render function

williaster commented on superset/assets/javascripts/explore/components/controls/ViewportControl.jsx:
```diff
@@ -0,0 +1,92 @@
+  }
+  renderPopover() {
+    return (
+      <Popover id="filter-popover" title="Viewport">
```
should `id` be unique in case there are multiple deck vis's?

mistercrunch replied:
react-boostrap made it required for accessibility, it should probably be unique anyways. I'm thinking the control name is probably sufficient since we don't have any multi-explore views at this time
<img width="892" alt="screen shot 2017-11-12 at 11 39 25 am" src="https://user-images.githubusercontent.com/487433/32702584-3579a3cc-c79e-11e7-8e25-fe4d45b183e7.png">

williaster commented on superset/assets/visualizations/deckgl/DeckGLContainer.jsx:
```diff
@@ -0,0 +1,91 @@
+  }
+  componentWillMount() {
+    const timer = setInterval(this.tick, 1000);
+    this.setState({ timer });
```
ditto functional `setState`

williaster commented on superset/assets/visualizations/deckgl/DeckGLContainer.jsx:
```diff
@@ -0,0 +1,91 @@
+    this.setState({ timer });
+  }
+  componentWillReceiveProps(nextProps) {
+    this.setState({
```
ditto functional `setState`

williaster commented on superset/assets/visualizations/deckgl/DeckGLContainer.jsx:
```diff
@@ -0,0 +1,91 @@
+    this.clearInterval(this.state.timer);
+  }
+  onViewportChange(viewport) {
+    const vp = Object.assign({}, viewport);
```
nit `{ ...spread }`, need a lint rule for this.

mistercrunch replied:
What about we do the spread change as a large linting PR across the board that would clean all Object.assign and set up the linting rule?

mistercrunch replied:
https://github.com/apache/incubator-superset/issues/3835

williaster commented on superset/assets/visualizations/deckgl/DeckGLContainer.jsx:
```diff
@@ -0,0 +1,91 @@
+      <MapGL
+        {...viewport}
+        mapStyle={this.props.mapStyle}
+        onViewportChange={this.onViewportChange.bind(this)}
```
shouldn't bind in a render

williaster commented on superset/assets/visualizations/deckgl/grid.jsx:
```diff
@@ -0,0 +1,43 @@
+  }));
+
+  const layer = new GridLayer({
+    id: 'screengrid-layer',
```
should this include the something to be unique?

williaster commented on superset/assets/visualizations/deckgl/scatter.jsx:
```diff
@@ -0,0 +1,56 @@
+    data,
+    pickable: true,
+    fp64: true,
+    // onHover: info => console.log('Hovered:', info),
```
remove?

williaster commented on superset/assets/visualizations/deckgl/scatter.jsx:
```diff
@@ -0,0 +1,56 @@
+  });
+
+  const layer = new ScatterplotLayer({
+    id: 'scatterplot-layer',
```
should this include the something to be unique?

williaster commented on superset/assets/visualizations/deckgl/hex.jsx:
```diff
@@ -0,0 +1,43 @@
+  }));
+
+  const layer = new HexagonLayer({
+    id: 'screengrid-layer',
```
should this include the something to be unique?

williaster commented on superset/assets/visualizations/deckgl/screengrid.jsx:
```diff
@@ -0,0 +1,43 @@
+  // Passing a layer creator function instead of a layer since the
+  // layer needs to be regenerated at each render
+  const layer = () => new ScreenGridLayer({
+    id: 'screengrid-layer',
```
should this include the something to be unique?

williaster reviewed (commented):
Super excited for this! Had a few minor comments 🚀

mistercrunch committed:
Linting

mistercrunch committed:
Addressing chri's comments

mistercrunch commented:
@williaster I addressed your comments except for `setState` where I think we should stick with the way it's now, and for the spread operator where I think it's out of scope for this PR but I think we should totally remove `Object.assign` from our code base.

graceguo-supercat commented:
LGTM
